google/cloud/firestore_v1/_helpers.py

```diff
@@ -1041,3 +1084,126 @@ def make_retry_timeout_kwargs(retry, timeout) -> dict:
         kwargs["timeout"] = timeout
 
     return kwargs
+
+
+def build_timestamp(
+    dt: Optional[Union[DatetimeWithNanoseconds, datetime.datetime]] = None
+) -> Timestamp:
+    """Returns the supplied datetime (or "now") as a Timestamp"""
+    return _datetime_to_pb_timestamp(dt or DatetimeWithNanoseconds.utcnow())
+
+
+def compare_timestamps(
+    ts1: Union[Timestamp, datetime.datetime], ts2: Union[Timestamp, datetime.datetime],
+) -> int:
+    dt1 = build_timestamp(ts1) if not isinstance(ts1, Timestamp) else ts1
+    dt2 = build_timestamp(ts2) if not isinstance(ts2, Timestamp) else ts2
+    dt1_nanos = dt1.nanos + dt1.seconds * 1e9
+    dt2_nanos = dt2.nanos + dt2.seconds * 1e9
+    if dt1_nanos == dt2_nanos:
+        return 0
+    return 1 if dt1_nanos > dt2_nanos else -1
+
+
+def deserialize_bundle(serialized: Union[str, bytes], client: "google.cloud.firestore_v1.client.BaseClient") -> "google.cloud.firestore_bundle.bundle.FirestoreBundle":  # type: ignore
+    """Inverse operation to a `FirestoreBundle` instance's `build()` method.
+    """
+    from google.cloud.firestore_bundle.bundle import FirestoreBundle
+    from google.cloud.firestore_bundle.types import BundleElement
+
+    # Outlines the legal transitions from one BundleElement to another.
+    bundle_state_machine = {
+        "__initial__": ["metadata"],
+        "metadata": ["named_query", "document_metadata", "__end__"],
+        "named_query": ["named_query", "document_metadata", "__end__"],
+        "document_metadata": ["document"],
+        "document": ["document_metadata", "__end__"],
+    }
+    allowed_next_element_types: List[str] = bundle_state_machine["__initial__"]
+
+    # This must be saved and added last, since we cache it to preserve timestamps,
+    # yet must flush it whenever a new document or query is added to a bundle.
+    # Thus, it must be the final step to deserialization.
+    metadata_bundle_element: Optional[BundleElement] = None
+
+    bundle: Optional[FirestoreBundle] = None
+    data: Dict
+    for data in _parse_bundle_elements_data(serialized):
+
+        # BundleElements are serialized as JSON containing one key outlining
+        # the type, with all further data nested under that key
+        keys: List[str] = list(data.keys())
+
+        if len(keys) != 1:
+            raise ValueError("Expected serialized BundleElement with one top-level key")
+
+        key: str = keys[0]
+
+        if key not in allowed_next_element_types:
+            raise ValueError(
+                f"Encountered BundleElement of type {key}. "
+                f"Expected one of {allowed_next_element_types}"
+            )
+
+        # Create and add our BundleElement
+        bundle_element: BundleElement = BundleElement.from_json(json.dumps(data))  # type: ignore
```

Comment:
Is this the only way to do this? it seems odd to me to serialize and deserialize again?

Reply:
I had expected it to be simpler, too, but this was all I found that worked. The form that seems like it would work, (`BundleElement(**data)`) errors out when it encounters Timestamps.